MBS-7043: Add basic touch support

sh-csg opened the pull request.

PhMemmel commented on amd/src/learningmap.js:
```diff
@@ -312,15 +317,57 @@ export const init = () => {
+                evt.preventDefault();
+            }
             if (evt.target.classList.contains('learningmap-draggable')) {
+                if (evt.type == 'touchstart') {
```
It would reduce complexity and improve readability if the touch events could be handled by separate event listener functions (could be added in the block at line 222 ff). These functions could do their touch stuff and after that call the "normal" event handler.

sh-csg replied:
Done

PhMemmel commented on amd/src/learningmap.js:
```diff
@@ -515,7 +573,7 @@ export const init = () => {
         event.preventDefault();
         hideContextMenu();
-        if (event.target.classList.contains('learningmap-place') && selectedElement === null) {
+        if (event.target.classList.contains('learningmap-place') && (selectedElement === null || event.type == 'touchend')) {
```
If going for the separate handling of touch events: Could this be realized by setting `selectedElement` to `null` in case of receiving a touchend event? Would make this a bit simpler.

sh-csg replied:
Done

PhMemmel replied:
That's still present?

PhMemmel reviewed (commented):
Could only test with dev tools mobile device simulation, looks good. The time I need to wait for the modal to popup is a bit long, maybe this can be reduced a bit. I also added a suggestion to refactor the handling of touch events.

sh-csg commented on amd/src/learningmap.js:
```diff
@@ -312,15 +317,57 @@ export const init = () => {
+                                    showContextMenu(evt);
+                                }
+                            },
+                            2000,
```
I shortened the timeout - but this might interfere with dragging the places. Maybe we have to wait for user feedback on this.

sh-csg replied:
Unfortunately a shorter timeout interferes a lot with dragging and double clicking.

PhMemmel commented on amd/src/learningmap.js:
```diff
@@ -373,11 +383,78 @@ export const init = () => {
+                            }
+                        },
+                        2000,
+                        evt
```
is this additional param needed?

sh-csg replied:
To hand over the event to the called function as a parameter

PhMemmel replied:
Yes, I know what the param means, I just wanted to ask, if it's needed there. But I tested (mainly out of personal interest ;-)) and you're totally right. It has to be passed there to enable the arrow function to pass it onto `showContextMenu`. Sorry for asking :-)

PhMemmel commented on amd/src/learningmap.js:
```diff
@@ -373,11 +383,78 @@ export const init = () => {
+                        () => {
+                            touchstart = false;
+                        },
+                    300 );
```
remove space?

PhMemmel commented on amd/src/learningmap.js:
```diff
@@ -373,11 +383,78 @@ export const init = () => {
+                        () => {
+                            touchstart = false;
+                        },
+                    300 );
```
remove space?

PhMemmel commented on amd/src/learningmap.js:
```diff
@@ -373,11 +383,78 @@ export const init = () => {
+         * @param {*} evt
+         */
+        function endTouch(evt) {
+            if (touchmove < 3 && touchstart) {
```
I think a comment explaining what `touchmove < 3` is for would be nice

PhMemmel commented on amd/src/learningmap.js:
```diff
@@ -22,6 +22,11 @@ export const init = () => {
 
+    // Variables for simulating double click on touch devices
+    var touchstart = false;
+    var touchmove = 0;
```
A comment would be nice explaining what this is for

PhMemmel commented on amd/src/learningmap.js:
```diff
@@ -330,7 +337,10 @@ export const init = () => {
+            if (evt.cancelable) {
+                evt.preventDefault();
+            }
+            touchmove++;
```
A comment why this is being incremented would be nice

sh-csg commented:
All requested changes are done, branch is rebased on latest master. Moving version number to 0.8.

sh-csg commented:
Also changed supported moodle versions (due to 4.1 and as a separate branch for 3.9/3.10 needs to be created).

PhMemmel commented on amd/src/learningmap.js:
```diff
@@ -285,11 +292,11 @@ export const init = () => {
-            el.addEventListener('touchcancel', endDrag);
+            el.addEventListener('touchend', endTouch);
+            el.addEventListener('touchleave', endTouch);
+            el.addEventListener('touchcancel', endTouch);
```
This is much more cleaner now, gj!

PhMemmel commented:
From code perspective it now looks good. I also tested again via browser dev tools, could not find any issues. This should be tested as soon as possible on a real mobile device. Just a minor thought:

The context menu is always being shown in the left upper corner now. Is this a mobile device optimization or is it intended to open it at the point where I "click" on a touch device?

And another thought which comes to my mind right now: Maybe one could circumvent the usage of the boolean variables to detect "fake touch double click" etc. by using promises. But that's just a quick thought, I do not have a precise prospect on how this would work. Thoughts? However, current code is clean, understandable and stable, so this definitely does not need to be implemented for this.

Good job!

sh-csg commented:
Changing the way of implementation may be a task for a later version - I can't handle that in the current PR. Positioning of the modal is fixed now.

PhMemmel commented:
Yes, as I wrote: Not intended for this PR. Will double check positioning later.

PhMemmel commented:
Unfortunately, the last commit makes the context menu appear outside the screen on some small devices or out of the editor area which needs to be fixed. If this is too complicated, I think we also can live with the version before where it always appears in the left upper corner.

PhMemmel commented:
I think what we want for the context menu would be a behavior like in this demo. Maybe the solution provided there could also be implemented: https://mionskowski.pl/posts/positioning-a-context-menu-using-pure-css/

sh-csg commented:
Thanks a lot for the suggestion. I changed the way of implementation in the last commit:
The context menu appearing also outside of the window remains as an issue (but one has to decide whether this is really an issue). But it stays always inside the viewport.
position: fixed leads to another (new) issue when scrolling as the context menu keeps position.

PhMemmel commented:
Looks good in general. However, on small devices when opening the context menu for a place which is pretty much at the right edge of the map there is an overflow. Did you mean that with 'outside of the window'? Can this be easily fixed?

![image](https://user-images.githubusercontent.com/65113153/198202059-b93dd3c3-29d7-457a-9ec7-0b8e74b213a7.png)

sh-csg commented:
I was not able to reproduce that - but maybe this small change helps

PhMemmel commented:
Tested again based on the last commit: It seems to solve all the issues with the positioning of the modal, good job!

Also tested basic functionality: Adding and removing places, moving places around, opening modal (positioning now is great, also in all edge cases I tested). Could not test on *real* mobile devices, but in Chrome Dev Tools everything looks good.
